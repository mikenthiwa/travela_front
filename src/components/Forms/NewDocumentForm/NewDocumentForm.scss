.add-file {
	height: 20px;
  width: 55px;
  color: #4F4F4F;
  font-family: "DIN Pro";
  font-size: 15px;
  line-height: 20px;
  margin-top: 20px;
  margin-bottom: 10px;
}


.document-input {
  box-sizing: border-box;
  width: 100%;

  &__input-container{
    height: 61px;
    border: 1px dashed #D4D4D4;
    border-radius: 2px;
    background-color: #F8F8F8;

    display: flex;

    &__prompts {
      display: grid;
      grid-template-columns: auto 1fr;
      align-items: center;
      grid-column-gap: 1em;
      margin-left: 15.5px;
      cursor: pointer;

      input[type=file] {
        display: none;
      }

      &__img{
        display: grid;
        align-items: center;
        justify-content: center;
        width: 40px;
        height: 40px;
        background-color: #E1E3EC;
        border-radius: 50%;

         img {
          width: 20px;
          height: 20px;
        }
      }

      &__text{
        width: max-content;
        display: inline-grid;
        justify-content: left;
        grid-gap: 0.4em;
        margin-top: 1em;
        align-items: center;
        grid-template-rows: 20px 10px;

         & > p {
          font-family: "DIN Pro";
          font-size: 14px;
          line-height: 18px;
          color: #999999;

          .upload-text {

            &--underline {
              text-decoration: underline;
            }
          }
        }

        .maximum-file-size{
          color:#4F4F4F!important;
          font-size: small;
          opacity: 0.4;
        }
      }
    }
  }

  .file-errors {
    color: #999999;
    font-size: 12px;
    display:flex;
    justify-content: space-between;
    margin: 5px 0 10px 0;
  }
}

.loading-icon {
  border: 2px solid #f3f3f3;
  border-top: 2px solid #385BD3;
  border-radius: 50%;
  display: inline-block;
  width: 10px;
  margin-right: 10px;
  height: 10px;
  animation: spin .5s linear infinite;
}

@import 'responsiveness';
