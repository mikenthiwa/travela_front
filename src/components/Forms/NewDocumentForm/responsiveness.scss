@media (max-width: 479px) {
  .modal.add-checklist-item hr {
    margin: 0 -60px 0 0 !important;
    width: initial;
    height: 1px;
    max-width: 479.6px;
  }
}
