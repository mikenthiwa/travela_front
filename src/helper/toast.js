import toast from 'toastr';
import cancelBtn from '../images/toast-cancel.svg';

toast.options = {
  progressBar: false,
  closeButton: true,
  preventDuplicates: true,
  positionClass: 'toast-top-center',
  closeHtml: `
    <button class="custom-close-btn">
      <img alt="close btn"/>
    </button>
  `,
};

export const successMessage = message =>  toast.success(message);
export const errorMessage = message =>  toast.error(message);


