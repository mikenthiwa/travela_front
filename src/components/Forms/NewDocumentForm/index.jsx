import React, { PureComponent } from 'react';
import {PropTypes} from 'prop-types';
import axios from 'axios';
import toast from 'toastr';
import { Input, FormContext } from '../FormsAPI';
import DocumentAPI from '../../../services/DocumentAPI';
import { errorMessage } from '../../../helper/toast';
import SubmitArea from './FormFieldSets/SubmitArea';
import documentUpload from '../../../images/icons/document-upload.svg';
import './NewDocumentForm.scss';

class NewDocumentForm extends PureComponent {
  constructor(props) {
    super(props);
    this.defaultState = {
      values: {
        name: '',
        file: ''
      },
      errors: {},
      hasBlankFields: true,
      isSubmitting: false,
      uploadProgress: 0,
    };
    this.state = {...this.defaultState};
  }

  validate = (field = 'name') => {
    let { values, errors } = this.state;
    [errors, values] = [{ ...errors }, { ...values }];
    let hasBlankFields = false;
    !values[field]
      ? (errors[field] = 'This field is required')
      : (errors[field] = '');

    hasBlankFields = Object.keys(values).some(key => !values[key]);
    const fileNameRegex = /(^[A-Za-z])(.+)$/;
    if(values[field] && !fileNameRegex.test(values[field])) {
      errors[field] = 'File name must start with an alphabet';
      // set to true to disable 'save' button
      hasBlankFields = true;
    }
    this.setState(prevState => {
      return { ...prevState, errors, hasBlankFields };
    });
    return !hasBlankFields;
  };

  handleCancel = () => {
    const { closeModal } = this.props;
    this.setState({ ...this.defaultState });
    closeModal();
  };

  handleUpload = (e) => {
    e.preventDefault();
    const file = e.target.files[0];
    if (!file) {
      return;
    }
    const documents = ['image/jpeg', 'image/png', 'application/pdf'];
    if (!documents.includes(file.type)) {
      return toast.error('Incorrect file type uploaded');
    }
    if (file.size> 10**7*2){
      return toast.error('Incorrect file size uploaded');
    }

    const fileReader = new FileReader();
    fileReader.onload = () => {
      this.setState(
        prevState =>
          ({hasBlankFields: false,
            values:
            {
              ...prevState.values,
              name: (file.name).replace(/\.[^/.]+$/, ''),
              file: fileReader.result
            },
          }),
        this.validate
      );
    };
    fileReader.readAsDataURL(file);
   
  };

  handleUploadProgress = e => this.setState({ uploadProgress: e.loaded/e.total});

  handleSubmit = async event => {
    const { values } = this.state;
    this.setState({ hasBlankFields: true, isSubmitting: true });
    const { user, createDocument } = this.props;
    event.preventDefault();
    const formData = new FormData();
    formData.append('file', values.file);
    formData.append('upload_preset', process.env.REACT_APP_PRESET_NAME);
    try {
      let cloudinaryUrl;
      let publicId;
      delete axios.defaults.headers.common['Authorization'];
      if (values.file !== '') {
        const uploadedDoc = await axios
          .post(process.env.REACT_APP_CLOUNDINARY_API, 
            formData, { onUploadProgress: this.handleUploadProgress});

        cloudinaryUrl = uploadedDoc.data.secure_url;
        publicId = uploadedDoc.data.public_id;
        this.setState({ isSubmitting: false, uploadProgress: 0 });
      }
      DocumentAPI.setToken();
      const documentData = {
        name: values.name,
        cloudinary_public_id: publicId,
        cloudinary_url: cloudinaryUrl,
        userId: user.result.userId
      };
      createDocument(documentData);

    } catch (error) {
      errorMessage('Unable to upload document');
      DocumentAPI.setToken();
      this.setState({ hasBlankFields: false, isSubmitting: false, uploadProgress: 0 });
    }
  };

  renderFileSelection = () => (
    <div className="document-input__input-container">
      <label className="document-input__input-container__prompts" htmlFor="select-file">
        <img
          src={documentUpload}
          alt="Document Upload" className="document-input__input-container__prompts__img" />
        <div className="document-input__input-container__prompts__text"><p>Upload File</p></div>
        <input type="file" onChange={this.handleUpload} id="select-file" />
      </label>
    </div>
  );

  render() {
    const { errors, values, hasBlankFields, isSubmitting, uploadProgress } = this.state;
    const { isUploading } = this.props;
    return (
      <FormContext
        targetForm={this} errors={errors} values={values}
        validatorName="validate"// uses default validator if this is undefined
      >
        <form onSubmit={this.handleSubmit}>
          <div className="add-file">Add File</div>
          <div className="document-input">
            { this.renderFileSelection() }
            { isSubmitting && <progress className="progress-bar" value={uploadProgress} max={1} /> }
            <div className="file-errors">
              <span>Allowed files: *jpeg, *pdf, *png</span>
              <span id="size">Max size: 10mb</span>
            </div>
          </div>
          <Input
            type="text"
            name="name"
            label="Name"
          />
          <SubmitArea
            hasBlankFields={hasBlankFields} onCancel={this.handleCancel}
            loading={isSubmitting || isUploading} />
        </form>
      </FormContext>
    );
  }
}

NewDocumentForm.propTypes = {
  closeModal: PropTypes.func.isRequired,
  isUploading: PropTypes.bool.isRequired,
};

export default NewDocumentForm;
