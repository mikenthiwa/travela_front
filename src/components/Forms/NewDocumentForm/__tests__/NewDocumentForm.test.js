import React from 'react';
import toast from 'toastr';
import NewDocumentForm from '..';
import moxios from 'moxios';

toast.error = jest.fn();
const realFileReader = FileReader;

const readAsDataURL = jest.fn(async () => {});
const mockFileReader = jest.fn(() => ({ readAsDataURL }));

describe('<NewDocumentForm />', () => {
  let wrapper;
  const props = {
    user: {
      result: {
        userId: 'user id',
      }
    },
    createDocument: jest.fn(),
    closeModal: jest.fn(),
    isUploading: false,
  };

  const file = new Blob(['This is a valid pdf file'], {type : 'application/pdf'});
  file.name = 'file.pdf';

  const invalidFileType = new Blob(['This is an invalid file type'], {type : 'text/plain'});
  invalidFileType.name = 'invalidFileType.txt';

  const invalidFileSize = new Blob(['x'.repeat(20000001)], {type : 'application/pdf'});
  invalidFileSize.name = 'invalidFileSize.pdf';

  const event = {
    preventDefault: jest.fn(),
    target: {
      files: [file]
    }
  };

  beforeEach(() => {
    wrapper = mount(<NewDocumentForm {...props} />);
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
    toast.error.mockReset();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the closeModal prop when click is simulated', () => {
    wrapper.find('#cancel').simulate('click');
    expect(props.closeModal).toHaveBeenCalledTimes(1);
  });

  it('should not crash when a file is not selected', async (done) => {
    event.target.files = [];
    expect(async () => wrapper.find('#select-file').simulate('change', event)).not.toThrow(TypeError);
    done();
  });

  it('should call the FileReader method when change is simulated', done => {
    event.target.files = [file];
    // use mock file reader to ensure the file reader is initialized correctly
    window.FileReader = mockFileReader;
    wrapper.find('#select-file').simulate('change', event);
    expect(readAsDataURL).toHaveBeenCalledWith(file);

    // use the real file reader so that onload is called
    window.FileReader = realFileReader;
    wrapper.find('#select-file').simulate('change', event);
    setTimeout(() => {
      expect(wrapper.state().values.name).toEqual(file.name.replace(/\.[^/.]+$/, ''));
      done();
    }, 100);
  });

  it('should not crash when a problem occurs during upload', async (done) => {
    //set a mock cloudinary URL
    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-cloudinary-api-fail';

    // set the state with data just enough for submit to be called
    await wrapper.setState({ ...wrapper.state(), values: {
      name: file.name,
      file,
    }});

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, { status: 500 });

    await wrapper.simulate('submit');
    moxios.wait(() => {
      expect(toast.error).toHaveBeenCalledWith('Unable to upload document');
      done();
    });
  });

  it('should upload to cloudinary then call createDocument on submit', async (done) => {
    // set the state with data just enough for submit to be called
    wrapper.setState({ ...wrapper.state(), values: {
      name: file.name,
      file,
    }});

    // set the mock cloudinary url to be called
    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-cloudinary-api-succeed';

    const cloudinaryResponse = {
      public_id: 'public id',
      secure_url: 'secure url',
    };

    // the expected response after uploading to cloudinary
    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 200,
      response: cloudinaryResponse
    });

    await wrapper.simulate('submit');
    moxios.wait(() => {
      expect(props.createDocument).toHaveBeenCalledWith({
        cloudinary_public_id: cloudinaryResponse.public_id,
        cloudinary_url: cloudinaryResponse.secure_url,
        name: file.name,
        userId: props.user.result.userId,
      });
      done();
    });
  });

  it('shows a toaster if file type is invalid', () => {
    event.target.files = [invalidFileType];
    wrapper.find('#select-file').simulate('change', event);
    expect(toast.error).toHaveBeenCalledWith('Incorrect file type uploaded');
  });

  it('shows a toast error if file size is greater than 20mb', () => {
    event.target.files = [invalidFileSize];
    wrapper.find('#select-file').simulate('change', event);
    expect(toast.error).toHaveBeenCalledWith('Incorrect file size uploaded');
  });

  it('shows progress bar while uploading', () => {
    wrapper.setState({ isSubmitting: true  });
    wrapper.instance().handleUploadProgress({ loaded: 1, total: 2});
    wrapper.update();
    const progressBar = wrapper.find('progress');
    expect(progressBar).toHaveLength(1);
    expect(progressBar.props().value).toEqual(0.5);
  });

  describe('NewDocumentForm.validate', () => {
    it('validates required field', () => {
      wrapper.instance().validate();
      expect(wrapper.state('errors')).toEqual({ name: 'This field is required' });
    });

    it('validates filename', () => {
      wrapper.state('values').name = '12filename'; // set name field value to string starting with digits
      wrapper.instance().validate();
      expect(wrapper.state('errors')).toEqual({
        name: 'File name must start with an alphabet'
      });
    });
  });

});
