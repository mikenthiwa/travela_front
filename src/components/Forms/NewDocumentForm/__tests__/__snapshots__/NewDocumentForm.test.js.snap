// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewDocumentForm /> renders correctly 1`] = `
<form
  key=".0"
  onSubmit={[Function]}
>
  <div
    className="add-file"
  >
    Add File
  </div>
  <div
    className="document-input"
  >
    <div
      className="document-input__input-container"
    >
      <label
        className="document-input__input-container__prompts"
        htmlFor="select-file"
      >
        <img
          alt="Document Upload"
          className="document-input__input-container__prompts__img"
          src="document-upload.svg"
        />
        <div
          className="document-input__input-container__prompts__text"
        >
          <p>
            Upload File
          </p>
        </div>
        <input
          id="select-file"
          onChange={[Function]}
          type="file"
        />
      </label>
    </div>
    <div
      className="file-errors"
    >
      <span>
        Allowed files: *jpeg, *pdf, *png
      </span>
      <span
        id="size"
      >
        Max size: 10mb
      </span>
    </div>
  </div>
  <div
    className="form-input  "
  >
    <label
      htmlFor="name"
    >
      Name
      <span
        style={
          Object {
            "color": "red",
          }
        }
      >
         * 
      </span>
    </label>
    <input
      className=" "
      label="Name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    <span
      className="error"
    />
  </div>
  <fieldset>
    <div
      className="submit-area"
    >
      <button
        className="bg-btn bg-btn--inactive"
        id="cancel"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="bg-btn bg-btn--active"
        disabled={true}
        id="submit"
        type="submit"
      >
        <undefined />
      </button>
    </div>
  </fieldset>
</form>
`;
